File: src/gestorAplicacion/Autocine/Autocine.java
package gestorAplicacion.Autocine;

import java.io.Serializable;
import java.util.*;
import java.util.stream.Collectors;
import gestorAplicacion.Taquilla.*;
import gestorAplicacion.Persona.Cliente;
import gestorAplicacion.Salas.Sala;
import gestorAplicacion.Salas.Sala2D;

/**
 * @author Jhon Ever Gallego Atehortua.
 * @param Clase Autocine
 * @summary Clase que contiene las salas, funciones, peliculas, clientes y metodos para la creación, modificación y observacion de los mismos.
 */

// Clase.
public class Autocine implements Serializable {
	
	// Serializacion.
	private static final long serialVersionUID = 1L;
	static List<Autocine> autocine;
	static {
		autocine = new ArrayList<Autocine>();
	}
	
	// Atributos.
	private String nombre;
	private static List<Cliente> clientes= new ArrayList<Cliente>();
	private static List<Funcion> cartelera= new ArrayList<Funcion>();
	private List<Pelicula> peliculas= new ArrayList<Pelicula>();
	private List<Sala> salas = new ArrayList<Sala>();
	
	// Constructores.
	public Autocine(String nombre) {
		this.nombre = nombre;
	}
	
	// Metodos.
	/**
	 * 
	 * @param mes
	 * @param dia
	 * @summary Recibe un mes y un dia.
	 * @return Una lista de salas disponibles.
	 */
	public ArrayList<Sala> salasDisponibles(int mes, int dia) {
		ArrayList<Sala> disponibles = new ArrayList<>();
		for(Sala sala: salas) { 
			if(sala.unoDisponible(dia, mes)) {
				disponibles.add(sala);
			}
		}
		return disponibles;
	}

	
	/**
	 * @param pelicula
	 * @param dia
	 * @param mes
	 * @summary Recibe una pelicula, un dia y mes.
	 * @return Una lista de funciones disponibles, que cumpla dichas condiciones.
	 */
	public ArrayList<Funcion> verFuncion(Pelicula pelicula, int dia, int mes) { 
		ArrayList<Funcion> funciones = new ArrayList<Funcion>();
		
		for(Funcion funcion: cartelera) { 
			if(funcion.getPelicula() == pelicula && funcion.getDia() >= dia && funcion.getMes() == mes ) {
				funciones.add(funcion);
			}
		}
		for(Funcion funcion: cartelera) { 
			if(funcion.getPelicula() == pelicula && funcion.getMes() > mes ){
				funciones.add(funcion);
			}
		}
		return funciones;
	}
	
	
	/**
	 * @param cliente
	 * @summary Recibe un cliente.
	 * @return Una lista de funciones disponibles segun el genero mas visto, genero seleccionado por el cliente.
	 */
	public ArrayList<Funcion> verFuncion(Cliente cliente) { 
		ArrayList<Funcion> funciones = new ArrayList<Funcion>();
		
		for(Funcion funcion: cartelera) {
			if(funcion.getPelicula().getGenero() == cliente.GeneroMasVisto()) {	
			}
		}
		return funciones;
	}
	
		
	/**
	 * @param dia
	 * @param mes
	 * @summary Recibe un dia y mes.
	 * @return Una lista de funciones disponibles segun el dia y mes seleccionado.
	 */
	public static ArrayList<Funcion> verFuncion( int dia, int mes) { 
		ArrayList<Funcion> funciones = new ArrayList<Funcion>(); 
		
		for(Funcion funcion: cartelera) {
			if(funcion.getDia() == dia && funcion.getMes() == mes){
				funciones.add(funcion);
			}
		}
		return funciones;
	}

	
	/**
	 * @param mes
	 * @summary Recibe un mes.
	 * @return Una lista de funciones disponibles segun el mes seleccionado.
	 */
	public ArrayList<Funcion> verFuncion(int mes) {	
		ArrayList<Funcion> funciones = new ArrayList<Funcion>(); 
		
		for(Funcion funcion: cartelera) {
			if(funcion.getMes() == mes) {
				funciones.add(funcion);
			}
		}
		return funciones;
	}
	

	/**
	 * @param boleto
	 * @summary Recibe un ticket.
	 * @return Una lista de funciones disponibles segun el ticket relacionado.
	 */
	public ArrayList<Funcion> verFuncion(Ticket ticket) {
		ArrayList<Funcion> funciones = new ArrayList<Funcion>(); 
	
		for(Funcion funcion: cartelera) {
			if(funcion == ticket.getFuncion()) {
				funciones.add(funcion);
			}
		}
		return funciones;
	}


	// Metodos para agregar elementos a las listas de la clase Autocine.
	/**
	 * @summary Metodos para agregar elementos a las listas de la clase Autocine.
	 */
	public void agregarCliente(Cliente nuevo) {
		clientes.add(nuevo);
	}
	public void agregarPelicula(Pelicula nuevo) {
		peliculas.add(nuevo);
	}
	public void agregarSala(Sala nuevo) {
		salas.add(nuevo);
	}
	public void agregarFuncion(Funcion nuevo) {
		cartelera.add(nuevo);
	}

	
	/**
	 * @param num
	 * @summary Recibe una cedula.
	 * @return Verifica si un cliente está en la lista de clientes.
	 */
	public static Boolean verificarCliente(int num) {
		ArrayList<Integer> lista = new ArrayList<Integer>();  
		
		for(Cliente cliente: clientes) {
			lista.add(cliente.getId());
		}
		
		if (lista.contains(num)) {	
			return true;
		}
		else {
			return false;
		}
		
	}
	
	
	/**
	 * @param num
	 * @summary Recibe una cedula.
	 * @return Retorna un objeto cliente cuya cedula concuerde con la ingresada.
	 */
	public static Cliente BuscadorCliente(int num) {
		ArrayList<Integer> lista = new ArrayList<Integer>();
		
		for(Cliente cliente: clientes) {
			lista.add(cliente.getId());
		
			if (cliente.getId() == num) {
				return cliente;
			}
		}
		return null ;
	}
	
	
	/**
	 * @param numero
	 * @summary Recibe un numero de funcion.
	 * @return Retorna un objeto funcion.
	 */
	public Funcion BuscadorFuncion(int numero) {
		ArrayList<Integer> lista = new ArrayList<Integer>();
		
		for(Funcion funcion: cartelera) {
			lista.add(funcion.getNumero());
			
			if(funcion.getNumero() == numero) {	
				return funcion;
			}
		}
		return null ;
	}
	
	
	/**
	 * @param num_silla
	 * @param funcion
	 * @summary Recibe un numero de puesto y una funcion.
	 * @return Un objeto de clase Ticket, cuyo numero de puesto sea igual al ingresado.
	 */
	public Ticket BuscadorTicket(int num_puesto, Funcion funcion) {
		ArrayList<Integer> lista= new ArrayList<Integer>();

		for(Ticket ticket: funcion.getTickets()) {
			lista.add(ticket.getNum_puesto());

			if(ticket.getNum_puesto() == num_puesto) {												
				return ticket;
			}
		}
		return null;
	}
	
	
	/**
	 * @param nombre
	 * @summary Recibe el nombre de una pelicula.
	 * @return Objeto de clase Pelicula, cuyo nombre coincida con el ingresado.
	 */
	public Pelicula BuscadorPelicula(String nombre) {
		ArrayList<String> lista = new ArrayList<String>();
		
		for(Funcion funcion: cartelera) {
			lista.add(funcion.getPelicula().getNombre());
			if (funcion.getPelicula().getNombre().equals(nombre)) {
				return funcion.getPelicula();
			}
		}
		return null ;
	}
	

	/**
	 * @param num
	 * @summary Recibe el numero de sala.
	 * @return Un objeto de clase Sala, cuyo numero coincida con el ingresado.
	 */
	public Sala buscarSala(int num) {
		List<Integer> lista = new ArrayList<Integer>();
		
		for(Sala sala: salas) {
			lista.add(sala.getNumero());
		}
		
		if (lista.contains(num)) {
			return salas.get(lista.indexOf(num));
		}
		else {
			return null;
		}
	}


	// Getters and Setters.
	public String getNombre() {
		return nombre;
	}
	public void setNombre(String nombre) {
		this.nombre = nombre;
	}
	public List<Cliente> getClientes() {
		return clientes;
	}
	public void setClientes(List<Cliente> clientes) {
		this.clientes = clientes;
	}
	public List<Funcion> getCartelera() {
		return cartelera;
	}
	public void setCartelera(List<Funcion> cartelera) {
		this.cartelera = cartelera;
	}
	public List<Pelicula> getPeliculas() {
		return peliculas;
	}
	public void setPeliculas(List<Pelicula> peliculas) {
		this.peliculas = peliculas;
	}
	public List<Sala> getSalas() {
		return salas;
	}
	public void setSalas(List<Sala> salas) {
		this.salas = salas;
	}
	public static List<Autocine> getAutocine() {
		return autocine;
	}
	public static void setAutocine(List<Autocine> autocine) {
		Autocine.autocine = autocine;
	}
	
	
}

File: Practica 2/AutoCine/src/gestorAplicacion/Autocine/Autocine.py
#**
#* @author: Jhon Ever Gallego Atehortua.
#* @param: Clase Autocine.
#* @summary: Clase Autocine. Clase que contiene as salas, funciones, peliculas, cientes y metodos para a creación, modificación y observación de los mismos.
#**

from gestorAplicacion.Taquilla.Funcion import Funcion
from gestorAplicacion.Taquilla.Pelicula import Pelicula
from gestorAplicacion.Taquilla.Funcion import Horario
from gestorAplicacion.Taquilla.Ticket import Ticket
from gestorAplicacion.Salas.Sala import Sala
from gestorAplicacion.Persona.Cliente import Cliente

# Clase.
class Autocine:
    
    # Atributos.
    def __init__(self, nombre):
        self._nombre = nombre
        self._clientes = []
        self._cartelera = []
        self._peliculas = []
        self._salas = []
       
        
    # Metodos.
    #**
	#* @param mes
	#* @param dia
	#* @summary Recibe un mes y un dia.
	#* @return Una lista de salas disponibles.
	#**
    def salasDisponibles(self, mes: int, dia: int):
        disponibles = list()
        
        for sala in self._salas:
            if Sala.unoDisponible(mes, dia):
                disponibles.append(sala)
                
        return disponibles
    
    
    #**
	#* @param pelicula
	#* @param dia
	#* @param mes
    #* @param cliente
	#* @summary Recibe una pelicula, un dia y mes.
	#* @return Una lista de funciones disponibles, que cumpla dichas condiciones.
	#**
    def verFuncion(self, *args) -> list:
        funciones: list = []

        if (len(args) == 3):
            pelicula: Pelicula = args[0]
            dia, mes = args[1:]
            for funcion in self.getCartelera():
                if(Funcion.getPelicula() == pelicula and Funcion.getDia() >= dia and Funcion.getMes() == mes):
                    funciones.append(funcion)
            for funcion in self.getCartelera():
                if(Funcion.getPelicula() == pelicula and Funcion.getMes() > mes):
                    funciones.append(funcion)

        # Este es el que recibe dia, mes.
        elif(len(args) == 2):
            dia, mes = args
            for funcion in self.getCartelera():
                if(Funcion.getDia() == dia and Funcion.getMes() == mes):
                    funciones.append(funcion)
        elif(len(args) == 1):
    
            # Recibe mes.
            if(type(args[0]) == int):
                mes = args[0]
                for funcion in self.getCartelera():
                    if(Funcion.getMes() == mes):
                        funciones.append(funcion)    
           
            # Recibe cliente. 
            else:
                cliente: Cliente = args[0]
                for funcion in self.getCartelera():
                   if(Funcion.getPelicula().getGenero() == cliente.GeneroMasVisto()):
                       funciones.append(funcion)
                       
        return funciones
    
    
    # Metodos para agregar elementos a las listas de la clase Autocine.
    #**
	#* @summary Metodos para agregar elementos a las listas de la clase Autocine.
	#**
    def agregarCliente(self, nuevo: Cliente):
       self._clientes.append(nuevo)

    def agregarPelicula(self, nuevo: Pelicula):
        self._peliculas.append(nuevo)

    def agregarSala(self, nuevo: Sala):
        self._salas.append(nuevo)

    def agregarFuncion(self, nuevo:Funcion):
        self._cartelera.append(nuevo)
    
    
    #**
	#* @param num
	#* @summary Recibe una cedula.
	#* @return Verifica si un cliente está en la lista de clientes.
	#**
    def verificarCliente(self, num: int) -> bool:
        lista : list = []
        
        for Cliente in self.getClientes():
            lista.append(Cliente.getId())
        
        return num in lista
    
    
    #**
	#* @param num
	#* @summary Recibe una cedula.
	#* @return Retorna un objeto cliente cuya cedula concuerde con la ingresada.
	#**
    def buscadorCliente(self, num : int):
        lista = self.getClientes()
        for cliente in lista:
            if (int(Cliente._id)== int(num)):
                return cliente
            
        return None
    
    
    #**
	#* @param num
	#* @summary Recibe el numero de sala.
	#* @return Un objeto de clase Sala, cuyo numero coincida con el ingresado.
	#**
    def buscarSala(self, num):
        lista = self.getSalas()
        for sala in lista:
            if (int(Sala.getNumero())== int(num)):
                return sala
            
        return None


    #**
	#* @param nombre
	#* @summary Recibe el nombre de una pelicula.
	#* @return Objeto de clase Pelicula, cuyo nombre coincida con el ingresado.
	#**
    def BuscadorPelicula(self, nombre):
        lista = self.getPeliculas()
        for pelicula in lista:
            if (str(Pelicula.getNombre())==nombre):
                return pelicula
            
        return None


    #**
	#* @param numero
	#* @summary Recibe un numero de funcion.
	#* @return Retorna un objeto funcion.
	#**
    def BuscadorFuncion(self,numero):
        lista=[]
        for funcion in self._cartelera:
            lista.append(Funcion.getNumero())
            if Funcion.getNumero()==int(numero): 
                return funcion
            
        return None


    #**
	#* @param num_silla
	#* @param funcion
	#* @summary Recibe un numero de puesto y una funcion.
	#* @return Un objeto de clase Ticket, cuyo numero de puesto sea igual al ingresado.
	#**
    def BuscadorTicket (self, num_puesto: int, funcion:Funcion):
        lista=[]
        for ticket in funcion.getTickets():
            lista.append(Ticket.getNum_puesto())

            if Ticket.getNum_puesto == num_puesto:  
                return ticket
            
        return None
    
    
    # Getters and Setters.
    def getNombre(self):
        return self._nombre
    def setNombre(self, nombre):
        self._nombre = nombre

    def getClientes(self):
        return self._clientes
    def setClientes(self, clientes):
        self._clientes = clientes

    def getCartelera(self):
      return self._cartelera
    def setCartelera(self, cartelera):
      self._cartelera = cartelera

    def getPeliculas(self):
     return self._peliculas
    def setPeliculas(self, peliculas):
        self._peliculas = peliculas

    def getSalas(self):
        return self._salas
    def setSalas(self, salas):
        self._salas = salas